making Bi/Tri/Sym times Diag output a structured matrix (not sparse)

mcognetta opened the pull request:
Fixes #31870 so that Bi/Tri/SymTriDiagonal matrices times Diagonal matrices output more structured types. Before, it outputted sparse matrices.

We also speed up multiplication in these cases.

-------------------

Some quick timings:

Pre-PR:
```Julia
D = Diagonal(rand(1000))
B = Bidiagonal(rand(1000), rand(999), 'U')
T = Tridiagonal(rand(1000, 1000))
S = SymTridiagonal(rand(1000), rand(999))

julia> @time B*D
  0.015049 seconds (28 allocations: 73.109 KiB)
1000×1000 SparseArrays.SparseMatrixCSC{Float64,Int64} with 1999 stored entries:

julia> @time D*B
  0.000131 seconds (8 allocations: 24.000 KiB)
1000×1000 Bidiagonal{Float64,Array{Float64,1}}:

julia> @time T*D
  0.011294 seconds (30 allocations: 137.234 KiB)
1000×1000 SparseArrays.SparseMatrixCSC{Float64,Int64} with 2998 stored entries:

julia> @time D*T
  0.010986 seconds (30 allocations: 137.234 KiB)
1000×1000 SparseArrays.SparseMatrixCSC{Float64,Int64} with 2998 stored entries:

julia> @time S*D
  0.011416 seconds (30 allocations: 137.234 KiB)
1000×1000 SparseArrays.SparseMatrixCSC{Float64,Int64} with 2998 stored entries:

julia> @time D*S
  0.011825 seconds (30 allocations: 137.234 KiB)
1000×1000 SparseArrays.SparseMatrixCSC{Float64,Int64} with 2998 stored entries:
```

Post-PR
```Julia
D = Diagonal(rand(1000))
B = Bidiagonal(rand(1000), rand(999), 'U')
T = Tridiagonal(rand(1000, 1000))
S = SymTridiagonal(rand(1000), rand(999))

julia> @time B*D
  0.000129 seconds (8 allocations: 24.000 KiB)
1000×1000 Bidiagonal{Float64,Array{Float64,1}}:

julia> @time D*B
  0.000036 seconds (8 allocations: 24.000 KiB)
1000×1000 Bidiagonal{Float64,Array{Float64,1}}:

julia> @time T*D
  0.000132 seconds (8 allocations: 24.016 KiB)
1000×1000 Tridiagonal{Float64,Array{Float64,1}}:

julia> @time D*T
  0.000040 seconds (8 allocations: 24.016 KiB)
1000×1000 Tridiagonal{Float64,Array{Float64,1}}:

julia> @time S*D
  0.000027 seconds (8 allocations: 24.016 KiB)
1000×1000 Tridiagonal{Float64,Array{Float64,1}}:

julia> @time D*S
  0.000032 seconds (8 allocations: 24.016 KiB)
1000×1000 Tridiagonal{Float64,Array{Float64,1}}:
```

mcognetta committed:
making Bi/Tri/Sym times Diag output a structured matrix (not sparse)

mcognetta committed:
fix test errors/typos

dkarrasch commented:
While you're at it, you could delete

https://github.com/JuliaLang/julia/blob/7e1eb6a9b0284e8f2e5fdc35331e46149ce25ee6/stdlib/LinearAlgebra/src/bidiag.jl#L500

It's not used, and equivalent to `BiTriSym`.

dkarrasch commented:
Oh, and what about multiplication with `Diagonal` from the left? It's seems like this is currently handled by some slow fallback as well?

mcognetta committed:
Merge branch 'master' of https://github.com/JuliaLang/julia into tri_times_diag

mcognetta committed:
adding diag*bi/tri/sym and removing unused union

mcognetta commented:
Thanks for the comments @dkarrasch .

I've updated the code so that `Diagonal*Bi/Tri/SymTridiagonal` uses a specialized multiply. I believe there are a few other places that we could achieve some speed ups or reduce the code complexity but maybe I will save it for a Juliacon Hackathon session?

dkarrasch commented:
Awesome, @mcognetta! Could you also add
```julia
mul!(C::AbstractMatrix, A::BiTriSym, B::Diagonal) = A_mul_B_td!(C, A, B)
mul!(C::AbstractMatrix, A::BiTriSym, B::Adjoint{<:Any,<:Diagonal}) = A_mul_B_td!(C, A, B)
mul!(C::AbstractMatrix, A::BiTriSym, B::Transpose{<:Any,<:Diagonal}) = A_mul_B_td!(C, A, B)
```
These would call the `A_mul_B_td!` method you wrote first.

Interestingly, we have methods for `Diagonal`, `Transpose{<:Any,<:Diagonal}`, `Adjoint{<:Any,<:Diagonal}`, but nothing for `Transpose{<:Any,<:Bidiagonal}` etc. I understand that if we follow this, then there is an explosion of possibilities. I think we should add methods for `transpose(::Bidiagonal)` that return `Bidiagonal` etc. also for complex number types, but that should probably be a different PR.

mcognetta commented:
My laptop just kicked the bucket so maybe someone else can add it so we can close this out. I'm not going to be able to push any code for several days.

dkarrasch committed:
add mul!(C,::BiTriSym, ::*Diagonal)

mcognetta committed:
Merge pull request #1 from dkarrasch/patch-1

add mul!(C,::BiTriSym, ::*Diagonal)

dkarrasch commented:
Hm, sorry, the `mul!` methods introduce ambiguities. I made another attempt, by splitting `AbstractVecOrMat`.

mcognetta commented:
Sorry, I can't test your PRs out. Did you try `detect_ambiguity`?

dkarrasch commented:
No I haven't, and I'm not sure how to do it locally. I know from other parts of the code that splitting helped. You could merge my PR and let CI test. If it doesn't work, revert the PRs, which would be bad, because then the specialized `A_mul_B!_td` wouldn't get called.

mcognetta committed:
Merge branch 'master' into tri_times_diag

mcognetta commented:
Finally got my computer to work on this. Seems like a big issue is coming from the fact that 
`B = Bidiagonal(rand(10), rand(9), 'U'); similar(B, Float64, (10, 10))` returns a `SparseMatrixCSC`.

I am going to dive into this today and see what I can get done.

mcognetta committed:
Merge remote-tracking branch 'origin' into tri_times_diag

mcognetta committed:
fixing ambiguious methods and adjoint/transpose

mcognetta committed:
space

mcognetta committed:
test error

changing `transpose`/`adjoint` to `Transpose`/`Adjoint`

This will fail until https://github.com/JuliaLang/julia/pull/31889 is merged.

mcognetta commented:
Bumping this.

I fixed the ambiguity issues and some of the other missing methods. This currently fails when `eltype` is BigFloat/BigInt since the `mul` methods can't get/set an `#undef` element. #32328 fixes this issue as a side effect, and all tests pass when that patch is applied to this PR.

dkarrasch commented on stdlib/LinearAlgebra/src/bidiag.jl:
```diff
@@ -437,6 +436,39 @@ function A_mul_B_td!(C::AbstractMatrix, A::BiTriSym, B::BiTriSym)
     C
 end
 
+function A_mul_B_td!(C::AbstractMatrix, A::BiTriSym, B::Diagonal)
```
Similarly to the comment below: do we need methods for `Transpose`s and `Adjoint`s of `B`? In line 634, you call `A_mul_B_td!` with a `B::Transpose{<:Any, <:Diagonal}`.

Hm, actually, it seems that the exact same code works for `B::Union{Diagonal, Adjoint{<:Any, <:Diagonal}, Transpose{<:Any, <:Diagonal}`. The elementwise transposition/adjoint is done at `getindex`. You could define a constant `AdjOrTransDiagonal` somewhere above?

mcognetta commented on stdlib/LinearAlgebra/src/bidiag.jl:
```diff
@@ -565,6 +629,11 @@ function *(A::SymTridiagonal, B::Diagonal)
+    TS = promote_op(matprod, eltype(A), eltype(B))
+    A_mul_B_td!(similar(A, TS), A, B)
+end
+
```
The thing is that even if we add all these `Transpose`/`Adjoint` methods, they will be slow due to `similar(Transpose(D), Float64)` returning a `SparseMatrix`. This is what I was alluding to in https://github.com/JuliaLang/julia/pull/31889#issuecomment-493701391.

Personally, I think that `similar` for structured matrices should always return the same type (unless the `dims` are not square), but this seems to be a design decision that a lot of thought went into.

mcognetta replied:
Perhaps this should be removed and put into another PR where we handle the Adjoint/Transpose case?

mcognetta committed:
remove transpose

mcognetta commented:
Ok, so I removed any of the `Transpose`/`Adjoint` changes that I added earlier so that we can move them to another PR since there are some other factors to consider. This PR now speeds up `Diagonal` times `Bi/Tri/SymTridiagonal` matrices on the left and the right. I have updated the PR description to reflect the new timings.

mcognetta commented:
Note that this is no longer dependent on #32328.

mcognetta commented:
Bump.
